[AMDGPU]: Rewrite mbcnt_lo/mbcnt_hi to work item ID where applicable

TejaX-Alaghari opened the pull request:
This PR aims to optimize `llvm.amdgcn.mbcnt.lo` and `llvm.amdgcn.mbcnt.hi` intrinsic patterns into simpler `workitem.id.x` operations when work group sizes are known at compile time for **improving performance of lane ID calculations**.

**visitMbcntLo:**
- Simple replacement: When `workgroup_size == wave_size` → Replace with `workitem.id.x`
- Bitmask optimization: When work group evenly splits into waves → Replace with `workitem.id.x & (wave_size - 1)`

**visitMbcntHi:**
- Copy optimization: On wave32, `mbcnt.hi(mask, val)` → `val` (upper 32 bits are always 0)
- Full pattern optimization: `mbcnt.hi(~0, mbcnt.lo(~0, 0))` → Replace with `workitem.id.x`


**Example 1: Simple Replacement**
```llvm
; Before:
%a = call i32 @llvm.amdgcn.mbcnt.lo(i32 -1, i32 0); With workgroup_size = 32
; After:
%a = call i32 @llvm.amdgcn.workitem.id.x()
```

**Example 2: Bitmask Optimization**
```llvm
; Before:
%a = call i32 @llvm.amdgcn.mbcnt.lo(i32 -1, i32 0); With workgroup_size = 64 (2 waves)
; After:
%tid = call i32 @llvm.amdgcn.workitem.id.x()
%a = and i32 %tid, 31
```

**Example 3: Copy Optimization**
```llvm
; Before:
%a = call i32 @llvm.amdgcn.mbcnt.hi(i32 -1, i32 %val); With workgroup_size = 32
ret i32 %a
; After:
ret i32 %val
```

**Example 4: Full Pattern Optimization**
```llvm
; Before:
%a = call i32 @llvm.amdgcn.mbcnt.lo(i32 -1, i32 0)
%b = call i32 @llvm.amdgcn.mbcnt.hi(i32 -1, i32 %a); With workgroup_size = 64
; After:
%b = call i32 @llvm.amdgcn.workitem.id.x()

; Before:
%a = call i32 @llvm.amdgcn.mbcnt.lo(i32 -1, i32 0)
%b = call i32 @llvm.amdgcn.mbcnt.hi(i32 -1, i32 %a); With workgroup_size = 48 (Partial mask)
; After:
%tid = call i32 @llvm.amdgcn.workitem.id.x()
%b = and i32 %tid, 31
```

chrisjbris commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1312,9 +1317,122 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+        break;
+
+      // hi arg0 must be all-ones
+      if (auto *HiArg0C = dyn_cast<ConstantInt>(II.getArgOperand(0))) {
```
```suggestion
      auto *HiArg0C = dyn_cast<ConstantInt>(II.getArgOperand(0));
      if(!HiArg0C || (HiArg0C && !HiArg0C->isAllOnesValue()))
          break;
```

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1312,9 +1317,122 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+          // Fallback: accept a declared function with the canonical name, but
+          // verify its signature to be safe: i32(i32,i32). Use the name
+          // comparison only when there's no intrinsic ID match.
+          if (CalledF->getName() == "llvm.amdgcn.mbcnt.lo") {
```
Do not match intrinsics by name, only use getIntrinsicID

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1312,9 +1317,122 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+          // verify its signature to be safe: i32(i32,i32). Use the name
+          // comparison only when there's no intrinsic ID match.
+          if (CalledF->getName() == "llvm.amdgcn.mbcnt.lo") {
+            if (FunctionType *FT = CalledF->getFunctionType()) {
```
You do not need to verify the signature

arsenm commented on llvm/test/Transforms/InstCombine/AMDGPU/mbcnt-to-bitmask-neg.ll:
```diff
@@ -0,0 +1,18 @@
+
+; ModuleID = 'mbcnt_to_bitmask_neg'
+
+define i32 @kernel() !reqd_work_group_size !1 {
```
```suggestion
```

arsenm commented on llvm/test/Transforms/InstCombine/AMDGPU/mbcnt-to-bitmask-neg.ll:
```diff
@@ -0,0 +1,18 @@
+; CHECK-NOT: and i32
+; CHECK-NOT: @llvm.amdgcn.workitem.id.x()
+
+; ModuleID = 'mbcnt_to_bitmask_neg'
```
```suggestion
```

TejaX-Alaghari replied:
Removed it

arsenm commented on llvm/test/Transforms/InstCombine/AMDGPU/mbcnt-to-bitmask-neg.ll:
```diff
@@ -0,0 +1,18 @@
+; RUN: opt -S -mtriple=amdgcn-unknown-amdhsa -passes=instcombine < %s | FileCheck %s
+; CHECK-NOT: and i32
+; CHECK-NOT: @llvm.amdgcn.workitem.id.x()
```
Generate full checks and avoid using CHECK-NOT

TejaX-Alaghari replied:
Done

arsenm commented on llvm/test/Transforms/InstCombine/AMDGPU/mbcnt-to-workitem-posit.ll:
```diff
@@ -0,0 +1,18 @@
+; CHECK-NOT: amdgcn.mbcnt_hi
+; CHECK: @llvm.amdgcn.workitem.id.x()
+
+define i32 @kernel() !reqd_work_group_size !0 {
```
This isn't a kernel, fix function name?

TejaX-Alaghari replied:
Renamed appropriately!

arsenm reviewed (commented):
A few issues that I think will prevent this from practically triggering:

- I'm not sure instcombine is the right place for this. Is there an instcombine run after AMDGPU attributor?
- Do we actually propagate reqd_work_group_size in attributor? It's an odd annotation since it's metadata like every other attribute, and I don't think we do (although we could)

TejaX-Alaghari commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1312,9 +1317,122 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+        // Fast-path: if this is a declared intrinsic, check the intrinsic ID.
+        if (CalledF->getIntrinsicID() == Intrinsic::amdgcn_mbcnt_lo) {
+          IsMbcntLo = true;
+        } else {
```
Done

TejaX-Alaghari commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -15,6 +15,7 @@
 //===----------------------------------------------------------------------===//
 
 #include "AMDGPUInstrInfo.h"
+#include "AMDGPUSubtarget.h"
```
Yes, you're right. Removed it now 👍

TejaX-Alaghari commented on llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:
```diff
@@ -1312,9 +1317,122 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
+        if (CalledF->getIntrinsicID() == Intrinsic::amdgcn_mbcnt_lo) {
+          IsMbcntLo = true;
+        } else {
+          // Fallback: accept a declared function with the canonical name, but
```
Initially, I had sketched to handle explicit calls to declared `@llvm.amdgcn.mbcnt.*` functions (not intrinsics). So call sites that are declared functions (not recognized IntrinsicInsts) are also handled.

But now I recognize its totally redundant! So, removed it now.

TejaX-Alaghari committed:
[AMDGPU] Move mbcnt optimization from InstCombine to AMDGPUCodeGenPrepare

This addresses reviewer concerns about pipeline timing by moving the mbcnt
optimization from InstCombine to AMDGPUCodeGenPrepare. The InstCombine pass
runs before AMDGPU Attributor, which means reqd_work_group_size metadata
may not be available. AMDGPUCodeGenPrepare runs later in the pipeline after
the attributor pass, ensuring proper metadata availability.

Changes:
- Move visitMbcntLo and visitMbcntHi methods to AMDGPUCodeGenPrepare
- Remove complex mbcnt optimization from AMDGPUInstCombineIntrinsic
- Keep simple wave32 mbcnt_hi -> copy optimization in InstCombine
- Move test files from InstCombine/AMDGPU to Transforms/AMDGPU
- Update test RUN lines to use amdgpu-codegenprepare pass

This fixes the pipeline ordering issue where InstCombine runs before
AMDGPU Attributor, preventing the optimization from triggering when
reqd_work_group_size metadata is set by the attributor.

krzysz00 commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -2113,6 +2119,147 @@ INITIALIZE_PASS_DEPENDENCY(UniformityInfoWrapperPass)
+  if (ST.isWaveSizeKnown())
+    Wave = ST.getWavefrontSize();
+
+  if (auto MaybeX = ST.getReqdWorkGroupSize(*F, 0)) {
```
I think we can still do the masking thing here? An x wavesize of 64 implies that the mbcnt is just `thread_id_x & 0x1f`?

krzysz00 reviewed (commented):
One note from me

TejaX-Alaghari committed:
[AMDGPU] Enhance mbcnt.lo optimization to support bitmask cases

Address reviewer feedback by extending visitMbcntLo to handle cases where
X dimension doesn't exactly match wave size but allows even wave distribution.

Key improvements:
- mbcnt.lo(~0, 0) -> workitem.id.x() & (wave_size-1) when X dimension
  allows wavefronts to evenly split using ST.hasWavefrontsEvenlySplittingXDim()
- Handles cases like X=48, X=64, X=128 on wave32 (previously unoptimized)
- Added comprehensive test coverage for bitmask optimization cases
- Updated existing test to reflect new correct behavior

This addresses krzysz00's comment about supporting masking optimizations
for non-exact wave size matches.

TejaX-Alaghari committed:
Fix mbcnt.hi wave32 optimization to respect reqd_work_group_size metadata

The wave32 optimization for mbcnt.hi was unconditionally optimizing the
instruction on wave32 targets, which caused CodeGen test failures because
the tests expected to see the mbcnt.hi instruction in the output.

This change makes the wave32 mbcnt.hi optimization conditional on the same
metadata requirements as the mbcnt.lo optimization, ensuring consistency
and preserving existing test behavior for functions without the appropriate
metadata.

krzysz00 commented:
> The AMDGPUCodegenPrepare pass is scheduled only for the higher optimization levels and it isn't included in the -O0 pipeline. Isn't this patch relevant for -O0 compilation?

I'd argue that is the sort of "optimization" that `-O0` can not do

TejaX-Alaghari committed:
Address review feedback: Add periods to all comments

As requested by @cdevadas, all comments in the mbcnt optimization
functions now end with a period for consistency with LLVM style.

Note: The use of 'auto' keyword was defended by @krzysz00 as these
appear next to dyn_cast<> which already includes the type information,
following common LLVM patterns.

arsenm commented:
> And I think all the `auto`s in the PR are legitimate - they're generally appearing next to `dyn_cast<>` that already include the type

There are ones that are not

TejaX-Alaghari commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -2113,6 +2119,181 @@ INITIALIZE_PASS_DEPENDENCY(UniformityInfoWrapperPass)
 INITIALIZE_PASS_END(AMDGPUCodeGenPrepare, DEBUG_TYPE, "AMDGPU IR optimizations",
                     false, false)
 
+bool AMDGPUCodeGenPrepareImpl::visitMbcntLo(IntrinsicInst &I) {
```
Changed the function signatures to make `visitMbcntLo` and `visitMbcntHi` methods const as requested as they don't modify the class state directly.

TejaX-Alaghari committed:
Address review comments from @arsenm

- Make visitMbcnt* functions const
- Use class member F instead of getFunction() calls
- Use auto* for pointer types and remove unnecessary null checks
- Remove range metadata from AND instruction (only apply to intrinsic calls)
- Simplify wave size logic by checking isWaveSizeKnown() upfront
- Use direct metadata access with F.getMetadata() instead of F->getMetadata()
- Add -mcpu=gfx906 to both test RUN lines to enable wave64 detection
- Update CHECK expectations to account for function attributes added
  when specifying a CPU target

All changes maintain the same optimization functionality while improving
code quality and following LLVM coding conventions.

TejaX-Alaghari commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -2113,6 +2119,181 @@ INITIALIZE_PASS_DEPENDENCY(UniformityInfoWrapperPass)
+    return false;
+
+  unsigned Wave = 0;
+  if (ST.isWaveSizeKnown())
```
Aborted if wave size is not known

TejaX-Alaghari committed:
Refactored test cases and implementation comments

TejaX-Alaghari committed:
Simplified the implementation per review suggestions -
1. Remove unnecessary casts
2. Replace manual AND instruction injection with ReplaceInstWithValue utility
3. Replace manual pattern matching with match utility
4. Rearange the logic to avoid redundancy
5. Add additional logic to check if the reqd wave size is a multiple of
Wave size for full pattern optimization

TejaX-Alaghari committed:
Addressed feedback on -
1. Add const qualifiers to mbcnt lo & hi visit functions
2. Reorganize partial wave size cases in the test files

TejaX-Alaghari commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -2113,6 +2119,172 @@ INITIALIZE_PASS_DEPENDENCY(UniformityInfoWrapperPass)
+        ST.makeLIDRangeMetadata(Tid);
+        IntegerType *ITy = cast<IntegerType>(Tid->getType());
+        Constant *Mask = ConstantInt::get(ITy, Wave - 1);
+        Instruction *AndInst = cast<Instruction>(B.CreateAnd(Tid, Mask));
```
Done

jmmartinez commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -2090,6 +2097,146 @@ INITIALIZE_PASS_DEPENDENCY(UniformityInfoWrapperPass)
+    // we can compute lane ID within wave using bit masking:
+    // lane_id = workitem.id.x & (wave_size - 1).
+    if (ST.hasWavefrontsEvenlySplittingXDim(F, /*RequiresUniformYZ=*/true)) {
+      if (isPowerOf2_32(Wave)) {
```
Isn't this always the case? (if we fear that may change maybe use an assert to signal that this code depends on that condition being true)

jmmartinez replied:
Sorry for the mix, I was talking about the `isPowerOf2_32(Wave)`

jmmartinez commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -2090,6 +2097,146 @@ INITIALIZE_PASS_DEPENDENCY(UniformityInfoWrapperPass)
+    // we can compute lane ID within wave using bit masking:
+    // lane_id = workitem.id.x & (wave_size - 1).
+    if (ST.hasWavefrontsEvenlySplittingXDim(F, /*RequiresUniformYZ=*/true)) {
+      if (isPowerOf2_32(Wave)) {
```
Same comment as before.

TejaX-Alaghari committed:
Remove redundant isPower2_32 check and combine 2 match functions in visitMbcntHi

TejaX-Alaghari committed:
Removed 4*Wave case from the tests

jmmartinez reviewed (approved):
Just a minor comment, the rest looks good to me.

TejaX-Alaghari committed:
Simplified negative cases test file as the arch specific cases are moved to a different file now

cdevadas commented on llvm/test/Transforms/AMDGPU/mbcnt-wave32-optimizations.ll:
```diff
@@ -0,0 +1,154 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 6
+; RUN: opt -S -mtriple amdgcn-unknown-amdhsa -mcpu=gfx1030 -passes=amdgpu-codegenprepare < %s | FileCheck %s
```
```suggestion
; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1030 -passes=amdgpu-codegenprepare < %s | FileCheck %s
```

cdevadas commented on llvm/test/Transforms/AMDGPU/mbcnt-wave64-optimizations.ll:
```diff
@@ -0,0 +1,95 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 6
+; RUN: opt -S -mtriple amdgcn-unknown-amdhsa -mcpu=gfx906 -passes=amdgpu-codegenprepare < %s | FileCheck %s
```
```suggestion
; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -mcpu=gfx906 -passes=amdgpu-codegenprepare < %s | FileCheck %s
```

cdevadas reviewed (approved):
LGTM with some nits.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -2090,6 +2097,139 @@ INITIALIZE_PASS_DEPENDENCY(UniformityInfoWrapperPass)
+    if (XLen == Wave) {
+      Function *WorkitemIdFn = Intrinsic::getOrInsertDeclaration(
+          I.getModule(), Intrinsic::amdgcn_workitem_id_x);
+      CallInst *NewCall = CallInst::Create(WorkitemIdFn, I.getName());
```
Would be shorter to use IRBuilder::CreateIntrinsic

TejaX-Alaghari replied:
Rewrote the code with `CreateIntrinsic`

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -2090,6 +2097,139 @@ INITIALIZE_PASS_DEPENDENCY(UniformityInfoWrapperPass)
+      // Construct optimized sequence: workitem.id.x & (wave_size - 1)
+      IRBuilder<> B(&I);
+      CallInst *Tid = B.CreateIntrinsic(Intrinsic::amdgcn_workitem_id_x, {});
+      ST.makeLIDRangeMetadata(Tid);
```
I guess this does do that underneath if the instruction is a call

TejaX-Alaghari commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -2090,6 +2097,139 @@ INITIALIZE_PASS_DEPENDENCY(UniformityInfoWrapperPass)
+    }
+  } else {
+    // When ST.getReqdWorkGroupSize() fails, use metadata. And only optimize the
+    // case when work group size = wave size.
```
Removed.

TejaX-Alaghari committed:
Address feedback about -
1. Use absolute path in headers inclusion
2. Eliminated redundant logic
3. Used CreateIntrinsic in place of getOrInsertDeclaration
4. Chanegd the triplet in test cases

TejaX-Alaghari commented on llvm/test/Transforms/AMDGPU/mbcnt-negative-cases.ll:
```diff
@@ -0,0 +1,131 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 6
+; RUN: opt -S -mtriple amdgcn-unknown-amdhsa -passes=amdgpu-codegenprepare < %s | FileCheck %s
```
Done

TejaX-Alaghari commented:
@arsenm, when you get some free time, can you please review the changes again and let me know if there's still further room for improvements!

TejaX-Alaghari committed:
Fixed format issue in AMDGPUCodeGenPrepare.cpp

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCodeGenPrepare.cpp:
```diff
@@ -2090,6 +2097,119 @@ INITIALIZE_PASS_DEPENDENCY(UniformityInfoWrapperPass)
+      Constant *Mask = ConstantInt::get(Tid->getType(), Wave - 1);
+      Value *AndInst = B.CreateAnd(Tid, Mask);
+      BasicBlock::iterator BI(&I);
+      ReplaceInstWithValue(BI, AndInst);
```
This same sequence is repeated in all of these cases, move to a helper function?

TejaX-Alaghari committed:
Refactor the code to remove redundant logic

arsenm reviewed (approved):
In some contexts we may want to go in the opposite direction. We could reclaim the work item ID register for something else (although that would be somewhat difficult to make useful

krzysz00 commented:
So, high-level ping on whether we should go ahead and land this
